teps/0044-decouple-task-composition-from-scheduling.md

```diff
@@ -0,0 +1,954 @@
+
+TBD - currently focusing on enumerating and examining alternatives before selecting one or more ways forward.
+
+## Alternatives
+
+Most of these options are not mutually exclusive:
+
+* [Task composition in Pipeline Tasks](#task-composition-in-pipeline-tasks)
+* [Update PipelineResources to use Tasks](#update-pipelineresources-to-use-tasks)
+* [Automatically combine Tasks based on workspace use](#automagically-combine-tasks-based-on-workspace-use)
+* [Introduce scheduling rules to Pipeline](#introduce-scheduling-rules-to-pipeline)
+* [PipelineRun: emptyDir](#pipelinerun-emptydir)
+* [Controller configuration](#controller-level)
+* [Within the Task](#within-the-task)
+* [Remove distinction between Tasks and Pipelines](#remove-distinction-between-tasks-and-pipelines)
+* [Custom Pipeline](#custom-pipeline)
+* [Create a new Grouping CRD](#create-a-new-grouping-crd)
+* [Rely on the Affinity Assistant](#rely-on-the-affinity-assistant)
+* [Custom scheduler](#custom-scheduler)
+* [Support other ways to share data (e.g. buckets)](#support-other-ways-to-share-data-eg-buckets)
+* [Focus on workspaces](#focus-on-workspaces)
+
+### Task composition in Pipeline Tasks
+
+In this option we make it possible to express Tasks which can be combined together to run sequentially as one pod.
+
+In the following example, 3 Tasks will be combined and run in one pod sequentially:
+
+1. `git-clone`
+2. `just-unit-test`
+3. `gcs-upload`
+
+```yaml
+apiVersion: tekton.dev/v1beta1
+kind: Pipeline
+metadata:
+  name: build-test-deploy
+spec:
+ params:
+  - name: url
+    value: https://github.com/tektoncd/pipeline.git
+  - name: revision
+    value: v0.11.3
+ workspaces:
+  - name: source-code
+  - name: test-results
+ tasks:
+ - name: run-unit-tests
+   taskRef:
+     name: just-unit-tests
+   workspaces:
+      - name: source-code
+      - name: test-results
+   init/before:
+      - taskRef: git-clone
+        params:
+       - name: url
+          value: $(params.url)
+        - name: revision
+          value: $(params.revision)
+        workspaces:
+        - name: source-code
+          workspace: source-code
+    finally/after:
+      - taskRef: gcs-upload
+        params:
+        - name: location
+          value: gs://my-test-results-bucket/testrun-$(taskRun.name)
+        workspaces:
+        - name: data
+          workspace: test-results
+```
+
+The `finally/after` Task(s) would run even if the previous steps fail.
+
+Pros:
+* Is an optional addition to the existing types (doesn't require massive re-architecting)
+* We have some initial indication (via PipelineResources) that this should be possible to do
+* Maintains a line between when to use a complex DAG and when to use this functionality since this is only sequential
+  (but the line is fuzzy)
+
+Cons:
+* Only helps us with some scheduling problems (e.g. doesn't help with parallel tasks or finally task execution)
+* What if you _don't_ want the last Tasks to run if the previous tasks fail?
```

Comment:
can the finally/after Tasks be guarded? what about the init Tasks?

Reply:
Also, how would these play with `when` guard ? It kinda fix the "I want to run this Task, no matter the previous task status", but not fully though.

Reply:
@jerop and @vdemeester - this is pretty much the first thing @ImJasonH said too, that people will very quickly want to be able to add more sophistication onto this - which is making me lean towards (in the long run) supporting the entire pipeline in a pod (cuz inevitably ppl will want all of the functionality) - but anyway i'll add this as a con!